core/server/provider.go

```diff
@@ -0,0 +1,107 @@
+package server
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"errors"
+	"io"
+	"time"
+
+	"golang.org/x/crypto/hkdf"
+)
+
+const (
+	keyValidDuration time.Duration = time.Hour * 24 * 3
+	keyRenewInterval time.Duration = time.Hour * 24
+)
+
+type Key struct {
+	Id    int
+	Value []byte
+	Start time.Time
+	End   time.Time
+}
+
+type Provider struct {
+	keys        map[int]Key
+	currentID   int
+	hkdf        io.Reader
+	lastCreated time.Time
+}
+
+func (k *Key) isValid() bool {
+	now := time.Now()
+	if !k.Start.Before(now) || !k.End.After(now) {
+		return false
+	}
+	return true
+}
+
+func (p *Provider) generateNext() (Key, error) {
+	p.currentID = p.currentID + 1
+	p.lastCreated = p.lastCreated.Add(keyRenewInterval)
+
+	key := Key{}
+	value := make([]byte, 32)
+	_, err := io.ReadFull(p.hkdf, value)
+	if err != nil {
+		return key, err
+	}
+
+	key.Value = value
+	key.Id = p.currentID
+	key.Start = p.lastCreated
+	key.End = key.Start.Add(keyValidDuration)
+
+	p.keys[p.currentID] = key
+
+	return key, nil
+}
+
+func (p *Provider) Init(id int) error {
+	p.currentID = id
+
```

Comment:
Why do we need a hkdf, when we seed it with a random value, anyways?. Couldn't we just generate the required number of cryptographically random bits and use them as a key?

Reply:
We seed it with the same random value for all servers. The hkdf would enable to not share the provider among go routines or even run the server on separate machines. Since after that the key generation uses hkdf.

Reply:
Done